lib/client/tncon/tncon.c

```diff
@@ -0,0 +1,883 @@
+* OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+* IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+* NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+* DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+* THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+* THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#include <stdlib.h>
+#include <stdio.h>
+#include <stddef.h>
+#include <windows.h>
+#include "ansiprsr.h"
+#include "tncon.h"
+
+#include "_cgo_export.h"
+
+// NOTE: Derived from https://github.com/PowerShell/openssh-portable/blob/0b73c4636d38f0c69424721d52d0e7752db99c81/contrib/win32/win32compat/tncon.c
+// Functions are largely kept unmodified, except where noted.
+
+// NOTE: vars that are normally set dynamically.
+//  - gbVTAppMode: some cursor functionality?
+//  - isAnsiParsingRequired: only needed if term doesn't support VT escapes
+//  - isConsoleVTSeqAvailable: if ENABLE_VIRTUAL_TERMINAL_INPUT is supported
+// extern bool gbVTAppMode;
+// extern BOOL isAnsiParsingRequired;
+// extern BOOL isConsoleVTSeqAvailable;
+
+// NOTE: Hardcoded defaults for the above vars. We ensure VT and ConPTY exist
+// before deferring to this C code. Hard-coding `isConsoleVTSeqAvailable`
+// implies the ENABLE_VIRTUAL_TERMINAL_INPUT flag is set; if we ever wish to
+// support older Windows releases, the flag can be switched to enable VT
+// sequence parsing.
+const bool gbVTAppMode = false;
+const BOOL isAnsiParsingRequired = FALSE;
+const BOOL isConsoleVTSeqAvailable = TRUE;
+
+char *glob_out = NULL;
+int glob_outlen = 0;
+int glob_space = 0;
+unsigned char  NAWSSTR[] = { "\xff\xfa\x1f\x00\x00\x00\x00\xff\xf0" };
+unsigned char tmp_buf[30];
+
+/* terminal global switches*/
+TelParams Parameters = {
+	0,		/* int fLogging */
+	NULL,		/* FILE *fplogfile */
+	NULL,		/* char *pInputFile */
+	NULL,		/* char *szDebugInputFile */
+	FALSE,		/* BOOL fDebugWait */
+	0,		/* int timeOut */
+	0,		/* int fLocalEcho */
+	0,		/* int fTreatLFasCRLF */
+	0,		/* int	fSendCROnly */
+	ENUM_LF,	/* int nReceiveCRLF */
+	'`',		/* char sleepChar */
+	'\035',		/* char menuChar; // CTRL-]  */
+	0,		/* SOCKET Socket */
+	FALSE,		/* BOOL bVT100Mode */
+	"\x01",		/* char *pAltKey */
+};
+
+TelParams* pParams = &Parameters;
+
+void GetVTSeqFromKeyStroke(INPUT_RECORD inputRecord);
+
+/* Write to a global buffer setup by ReadConsoleForTermEmul() */
+int
+WriteToBuffer(char* source, size_t len)
+{
+	// NOTE: small modification to send off an event here.
+	writeSequenceEvent(source, len);
+
+	while (len > 0) {
+		if (glob_outlen >= glob_space)
+			return glob_outlen;
+		*glob_out++ = *source++;
+		len--;
+		glob_outlen++;
+	}
+	return glob_outlen;
```

Comment:
if we only care about our Go callback, can all this global buffer business be removed?

Reply:
I've now removed the global buffer (and all other global state).